### PR TITLE
[OM] Implement tuple operations

### PR DESCRIPTION
**Pull request by uenoku**
This commit implements tuple_create/tuple_get operations

### CODE REVIEW

File: include/circt/Dialect/OM/OMOps.td [hunk 1]
```diff
@@ -195,4 +197,82 @@ def ListCreateOp : OMOp<"list_create", [Pure, SameTypeOperands]> {
...
+      ::mlir::MLIRContext *context, ::std::optional<::mlir::Location> location,
+      ::mlir::ValueRange operands, ::mlir::DictionaryAttr attributes,
+      ::mlir::OpaqueProperties,
+      ::mlir::RegionRange regions,
+      ::llvm::SmallVectorImpl<::mlir::Type> &inferredReturnTypes) {
+      ::llvm::SmallVector<::mlir::Type> types;
+      for (auto op: operands) 
+        types.push_back(op.getType());
+      inferredReturnTypes.push_back(mlir::TupleType::get(context, types));
+      return ::mlir::success();
```
Comment (prithayan) [inline include/circt/Dialect/OM/OMOps.td:231]: Maybe this goes in `OMOps.cpp`, but not sure of the convention.

**Review by prithayan — APPROVED** [review]
Looks good.

File: include/circt/Dialect/OM/OMOps.td [hunk 2]
```diff
@@ -195,4 +197,82 @@ def ListCreateOp : OMOp<"list_create", [Pure, SameTypeOperands]> {
...
+  );
+
+  let assemblyFormat = [{
+    $input `[` $index `]` `:` type($input) attr-dict
+  }];
+
+  let extraClassDeclaration = [{
+    // Implement InferTypeOpInterface.
+    static ::mlir::LogicalResult inferReturnTypes(
+      ::mlir::MLIRContext * context, ::std::optional<::mlir::Location> location,
```
Comment (dtzSiFive) [inline include/circt/Dialect/OM/OMOps.td:263]: Nit:
```suggestion
      ::mlir::MLIRContext *context, ::std::optional<::mlir::Location> location,
```

File: include/circt/Dialect/OM/OMOps.td [hunk 3]
```diff
@@ -195,4 +197,82 @@ def ListCreateOp : OMOp<"list_create", [Pure, SameTypeOperands]> {
...
+      ::mlir::ValueRange operands, ::mlir::DictionaryAttr attributes,
+      ::mlir::OpaqueProperties, ::mlir::RegionRange regions,
+      ::llvm::SmallVectorImpl<::mlir::Type> & inferredReturnTypes) {
```
Comment (dtzSiFive) [inline include/circt/Dialect/OM/OMOps.td:266]: Nit:
```suggestion
      ::llvm::SmallVectorImpl<::mlir::Type> &inferredReturnTypes) {
```

File: include/circt/Dialect/OM/OMOps.td [hunk 4]
```diff
@@ -195,4 +197,82 @@ def ListCreateOp : OMOp<"list_create", [Pure, SameTypeOperands]> {
...
+    auto idx = attributes.getAs<::mlir::IntegerAttr>("index");
+    inferredReturnTypes.push_back(
+        operands[0]
+            .getType()
+            .cast<mlir::TupleType>()
+            .getTypes()[idx.getValue().getLimitedValue()]);
```
Comment (dtzSiFive) [inline include/circt/Dialect/OM/OMOps.td:272]: Constraint or verification re:index being valid? (+tests for when this is wrong?)

File: include/circt/Dialect/OM/OMOps.td [hunk 5]
```diff
@@ -16,6 +16,8 @@
 include "circt/Dialect/OM/OMOpInterfaces.td"
 include "mlir/Interfaces/SideEffectInterfaces.td"
 include "mlir/IR/BuiltinAttributeInterfaces.td"
+include "mlir/Interfaces/InferTypeOpInterface.td"
```
Comment (dtzSiFive) [inline include/circt/Dialect/OM/OMOps.td:19]: Nit: Header sorting.  Same above in `OMOps.h`, but at least group headers under same directory together please.

**Review by dtzSiFive — APPROVED** [review]
Awesome, nice and general! Generally LGTM, added some comments.

File: include/circt/Dialect/OM/OMOps.td (hunk 4 above)
Comment (uenoku) [reply include/circt/Dialect/OM/OMOps.td]: Added verification and tests, thanks!

**Pull request merged**